File: src/tree.js
var Tree = function(value) {
  var newTree = Object.create(treeMethods);
  newTree.v = value;
  newTree.c = [];
  return newTree;
};

var treeMethods = {};

treeMethods.addChild = function(value) {
  var newChild = Tree(value);
  this.c.push(newChild);
  return newChild;
};

treeMethods.contains = function(target) {
  var found;
  if (arguments[1] === undefined) {
    found = false;
  } else {
    found = arguments[1];
  }
  // Look over the children
  this.c.forEach(function(child) {
    // With each child
      // If value matches
    if (child.v === target) {
      found = true;
    } else {
      // If doesn't; recurse
      found = child.contains(target, found);
    }
  });
  // Otherwise return false
  return found;
};

module.exports = Tree;

File: go/tree/tree.go
package tree

import (
	"bytes"
	"fmt"
)

// Tree struct
type Tree struct {
	Value    []byte
	Children []Tree
}

// Add new child node to tree
func (t *Tree) Add(v []byte) *Tree {
	c := Tree{Value: v}
	t.Children = append(t.Children, c)
	return t
}

// Contains return true if tree contains target
func (t *Tree) Contains(v []byte) bool {
	if v != nil {
		return false
	}

	if bytes.Compare(t.Value, v) == 0 {
		return true
	}

	for _, c := range t.Children {
		return c.Contains(v)
	}
	return false
}

// Traverse will print tree values
func (t *Tree) Traverse() {
	fmt.Println(t.Value)
	for _, c := range t.Children {
		c.Traverse()
	}
}
